accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,500 @@
+
+- A potential future possibility is mentioned under the `Specify the authentication plugin in NuGet.Config file` section.
+In addition to that, if we ever plan to provide users with an option to manage their plugins per repository instead of loading all the available plugins, we can consider using [.NET Local tools](https://learn.microsoft.com/dotnet/core/tools/local-tools-how-to-use).
+- The advantage of local tools is that the .NET CLI uses manifest files to keep track of tools that are installed locally to a directory.
+When the manifest file is saved in the root directory of a source code repository, a contributor can clone the repository and invoke a single .NET CLI command such as [`dotnet tool restore`](https://learn.microsoft.com/dotnet/core/tools/dotnet-tool-restore) to install all of the tools listed in the manifest file.
+- Having NuGet plugins under the repository folder eliminates the need for NuGet to load plugins from the user profile.
+
+### Support for NuGet plugins developed in non-.NET languages
+
+- The NuGet Client plugin code needs to be updated to utilize a standard RPC mechanism, such as StreamJsonRpc, in order to support NuGet plugins developed in languages other than .NET.
+The IPC (Inter-Process Communication) used by NuGet is custom-made, and it would be beneficial for plugin implementers if it were based on industry standards.
+This would allow plugin authors to reuse existing implementations for other languages.
+While there is no technical barrier preventing the implementation of a client in a non-.NET language, there is an additional cost associated with reimplementing the plugin protocol.
+This cost serves as a deterrent for writing non-.NET plugins.
+
+### Improve the discoverability of NuGet plugins published as .NET Tools
+
+- At present, the `dotnet pack` command disregards the `PackageType` property when the `PackAsTool` property is set to true.
+This might be because the new package type, `CredentialProvider`, which I added to the .csproj file, is not a recognized [`PackageType`](https://github.com/NuGet/NuGet.Client/blob/cecbc9a1f7a5cd0ea0a62dea2523f740bbd078d3/src/NuGet.Core/NuGet.Packaging/Core/PackageType.cs#L15-L20).
+If the `dotnet pack` command could generate a .nupkg for .NET Tool with multiple package types, we could introduce a new `dotnet nuget plugin search` command.
+This command would act as a wrapper for the `dotnet tool search` command, further refining the results based on the additional package type, such as `CredentialProvider`.
+These package types can be found in the `.nuspec` metadata file of the generated nupkg.
+
+### Support for other extensions on Windows
+
+- In the future, we could consider supporting extensions other than `.exe` configured in `PATHEXT` as executables.
+To achieve this, NuGet would need to identify the correct executable or interpretter to run a particular file.
+For example, to execute a PowerShell script, NuGet would need to launch the process as shown below.
```

Comment:
powershell is not a good example, because `.PS1` is not in default PATHEXT, and therefore you can't shell execute `script.ps1`.

Reply:
If I open a command prompt (cmd.exe, not powershell), this is what I get:

```console
C:\src>assoc .js
.js=JSFile

C:\src>ftype JSFile
JSFile=C:\Windows\System32\WScript.exe "%1" %*
```

So, `FileName` should be `C:\Windows\System32.WScript.exe`, not `node.exe`. Well, unless installing NodeJS changes the .js file association, which kind of proves the point that if we were to handle PATHEXT, we have to go to a lot of effort to do it "properly". Especially since NuGet invokes programs as console apps, so needs to use `CScript.exe` rather than `WScript.exe`.

Luckily for us, .bat files make it the cred provider author's responsibility for invoking it correctly.